File: go/145.go
/**
题目概述:
给定一个二叉树，返回它的 后序 遍历。

示例:

输入: [1,null,2,3]
   1
    \
     2
    /
   3

输出: [3,2,1]
进阶: 递归算法很简单，你可以通过迭代算法完成吗？


*/
package main


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 递归
func postorderTraversal(root *TreeNode) []int {
   var res []int
   var inner func(node *TreeNode)
   inner= func(node *TreeNode) {
	   if node ==nil{
		   return
	   }
	   inner(node.Left)
	   inner(node.Right)
	   res=append(res,node.Val)
   }
   inner(root)
   return res
}

// 非递归（栈,迭代）
func postorderTraversal(root *TreeNode) []int {
    var res []int
    var prev *TreeNode
    stack :=[]*TreeNode{}
    for root !=nil || len(stack)>0{
       for root !=nil{
       	 stack=append(stack,root)
       	 root=root.Left
	   }
       top:=len(stack)-1
       root=stack[top]
       stack=stack[:top]
       if root.Right ==nil || root.Right==prev{
       	  // 赋值
       	  res=append(res,root.Val)
       	  prev=root
       	  root=nil
	   }else{
	   	  stack=append(stack,root)
	   	  root=root.Right
	   }
	}
    return res
}

File: php/145.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    function postorderTraversal($root) {
      $res=[];
      $this->helper($root,$res);
      return $res;
    }

// 递归
    function helper($root,&$res){
      if($root==null)return;
      $this->helper($root->left,$res);
      $this->helper($root->right,$res);
      $res[]=$root->val;
    }

    // 迭代
     function postorderTraversal($root) {
          $res=[];
          $stack=[];
          $prev=null;
          while($root !=null || !empty($stack)){
             while($root !=null){
                $stack[]=$root;
                $root=$root->left;
             }
             $root=array_pop($stack);
             if($root->right ==null || $root->right ==$prev){
                $res[]=$root->val;
                $prev=$root;
                $root=null;
             }else{
                $stack[]=$root;
                $root=$root->right;
             }
          }
          return $res;
     }

}